File: cpp-test/utils/rand_bench.py
#!/usr/bin/env python3
import time
import argparse
import numpy as np
import caspian

def run_test(inputs, outputs, hidden, runs, seed, runtime):
    n_neurons = inputs + outputs + hidden

    n_input_synapses = min(hidden, 64)
    n_output_synapses = min(hidden, 64)
    n_hidden_synapses = min(hidden, 16)
    n_hidden_synapses_max = n_hidden_synapses * 2

    ####

    trnet_s = time.perf_counter()

    net = caspian.Network(n_neurons)
    net.make_random(n_inputs = inputs, 
                    n_outputs = outputs, 
                    seed = seed,
                    n_input_synapses = n_input_synapses,
                    n_output_synapses = n_output_synapses,
                    n_hidden_synapses = n_hidden_synapses,
                    n_hidden_synapses_max = n_hidden_synapses_max,
                    threshold_range = (0, 255),
                    weight_range = (0, 255),
                    delay_range = (0, 15),
                    leak_range = (0, 3))

    trnet_e = time.perf_counter()

    ####

    tcfg_s = time.perf_counter()
    
    sim = caspian.Simulator()
    sim.configure(net)

    for i in range(outputs):
        sim.monitor_output(i, 0, -1)

    tcfg_e = time.perf_counter()

    print("Seed: {} | Inputs: {} Outputs: {} | Neurons: {} Synapses: {} | Cycles: {}".format(
        seed, inputs, outputs, net.num_neurons(), net.num_synapses(), runtime))

    print("Random Net: {:.1f} us".format((trnet_e - trnet_s) * 10e6))
    print("Configure: {:.1f} us".format((tcfg_e - tcfg_s) * 10e6))

    # track simulation times
    sim_times = list()

    # do specified number of simulation runs
    for run in range(runs):

        # Apply inputs
        for i in range(inputs):
            sim.apply_input(i, 500, i)

        tsim_s = time.perf_counter()
        sim.simulate(runtime)
        accumulations = sim.get_metric("accumulate_count")
        tsim_e = time.perf_counter()

        sim_time = tsim_e - tsim_s

        print("Simulate {:3}: {:.9f} s".format(run, sim_time))

        sim_times.append(sim_time)
        sim.clear_activity()

    sim_times.sort()
    avg = np.sum(sim_times) / runs

    print("Average Simulate (s)     : {}".format(avg))
    print("Median Simulate (s)      : {}".format(np.median(sim_times)))
    print("Accumulations            : {}".format(accumulations))
    print("Accumulations per second : {:.1f}".format(accumulations / avg))
    print("Effective Clock Speed    : {:.4f} KHz".format( (runtime / avg) / 1000 ))

parser = argparse.ArgumentParser(description='Random Network Benchmark')
parser.add_argument('inputs',  metavar='I', type=int, help='number of input neurons')
parser.add_argument('outputs', metavar='O', type=int, help='number of output neurons')
parser.add_argument('hidden',  metavar='H', type=int, help='number of hidden neurons')
parser.add_argument('runs',    metavar='R', type=int, help='number of runs to complete')
parser.add_argument('runtime', metavar='T', type=int, help='number of cycles to simulate per run')
parser.add_argument('seed',    metavar='S', type=int, help='RNG seed value')

args = parser.parse_args()

run_test(inputs = args.inputs,
         outputs = args.outputs,
         hidden = args.hidden,
         runs = args.runs,
         runtime = args.runtime,
         seed = args.seed)

File: cpp-test/utils/rand_bench.cpp
#include "network.hpp"
#include "simulator.hpp"
#include <iostream>
#include <vector>
#include <chrono>

using namespace caspian;

void run_test(int inputs, int outputs, int hidden, int runs, int seed, int runtime = 0)
{
    int n_neurons = inputs + outputs + hidden;
    uint64_t accumulations = 0;

    std::vector<std::chrono::duration<double>> sim_times;

    int n_input_synapses = std::min(hidden, 64);
    int n_output_synapses = std::min(hidden, 64);
    int n_hidden_synapses = std::min(hidden, 16);
    int n_hidden_synapses_max = n_hidden_synapses * 2;

    auto rand_start = std::chrono::system_clock::now();

    Network net(n_neurons);

    // Generate the pass network
    net.make_random(inputs, outputs, seed, 
                    n_input_synapses,
                    n_output_synapses,
                    n_hidden_synapses,
                    n_hidden_synapses_max);

    // Configure the simulator with the new network
    auto cfg_start = std::chrono::system_clock::now();

    Simulator sim;
    sim.configure(&net);

    auto cfg_end = std::chrono::system_clock::now();

    std::chrono::duration<double, std::micro> rnd_duration = (cfg_start - rand_start);
    std::chrono::duration<double, std::micro> cfg_duration = (cfg_end - cfg_start);

    printf("Seed: %d | Inputs: %d Outputs: %d | Neurons: %zu Synapses: %zu | Cycles: %d\n", 
            seed, inputs, outputs, net.num_neurons(), net.num_synapses(), runtime);
    printf("Random Net: %lf us\n", rnd_duration.count()); 
    printf("Configure : %lf us\n", cfg_duration.count());

    for(int r = 0; r < runs; ++r)
    {
        auto sim_start = std::chrono::system_clock::now();

        // Queue up inputs
        for(int i = 0; i < inputs; ++i)
        {
            sim.apply_input(i, 500, i);
        }

        // Simulate with sufficient time (intentionally extra)
        sim.simulate(runtime);
        accumulations = sim.get_metric("accumulate_count");
        auto sim_end = std::chrono::system_clock::now();

        std::chrono::duration<double> sim_time = sim_end - sim_start;
        std::cout << "Simulate " << r << ": " << (sim_time).count() << " s" << std::endl;
        sim_times.push_back(sim_time);

        sim.clear_activity();
    }

    std::sort(sim_times.begin(), sim_times.end());

    double avg = 0;
    for(auto const &t : sim_times) avg += t.count();
    avg /= sim_times.size();
    printf("Average Simulate (s)     : %lf\n", avg);
    printf("Median Simulate  (s)     : %lf\n", sim_times[sim_times.size()/2].count());
    printf("Accumulations            : %llu\n", accumulations);
    printf("Accumulations per second : %.1f\n", static_cast<double>(accumulations) / avg);
    printf("Effective Clock Speed    : %.4f KHz\n", (static_cast<double>(runtime) / avg) / (1000) );
}

int main(int argc, char **argv)
{
    int inputs, outputs, hidden, runs, rt, seed;

    if(argc < 7)
    {
        printf("Usage: %s inputs outputs hidden n_runs runtime seed\n", argv[0]);
        exit(1);
    }

    inputs = atoi(argv[1]);
    outputs = atoi(argv[2]);
    hidden = atoi(argv[3]);
    runs = atoi(argv[4]);
    rt = atoi(argv[5]);
    seed = atoi(argv[6]);

    run_test(inputs, outputs, hidden, runs, seed, rt);
    return 0;
}

/* vim: set shiftwidth=4 tabstop=4 softtabstop=4 expandtab: */
